libs/core/src/main/java/org/opensearch/core/common/io/stream/StreamInput.java

```diff
@@ -632,6 +632,76 @@ public String[] readOptionalStringArray() throws IOException {
         return null;
     }
 
+    /**
+     * Reads an ordered {@link Map} from a data source using provided key and value readers.
+     *
+     * This method creates a `LinkedHashMap`, ensuring that the iteration order of the map
+     * matches the order in which the entries were read from the data source. It uses the
+     * `keyReader` and `valueReader` to read each key-value pair and constructs the map
+     * with an initial capacity optimized based on the expected size.
+     *
+     * Usage example:
+     * <pre><code>
+     * Map&lt;Integer, String&gt; orderedMap = readOrderedMap(StreamInput::readInteger, StreamInput::readString);
+     * </code></pre>
+     *
+     * @param keyReader The {@link Writeable.Reader} used to read the keys of the map.
+     * @param valueReader The {@link Writeable.Reader} used to read the values of the map.
+     * @return A {@link LinkedHashMap} containing the keys and values read from the data source.
+     *         The map maintains the order in which keys and values were read.
+     * @throws IOException If an I/O error occurs during reading from the data source.
+     */
+    public <K, V> Map<K, V> readOrderedMap(Writeable.Reader<K> keyReader, Writeable.Reader<V> valueReader) throws IOException {
```

Comment:
I think we should not have any reliance on order of maps / lists, the caller could use `readMap` and apply whatever ordering (sorting) it wants.

Reply:
its more of a convenient method to avoid reordering, and looks a bit cleaner to me, i'd suppose i can remove it but dont really have a super strong opinion. WDYT?

Reply:
I would better remove it, the convenience for one-off operation is not sufficient to make it into public API (in my opinion), thank you

Reply:
removed